### PR TITLE
Chore: remove hardcoded go version

### PR DESCRIPTION
**Pull request by ying-jeanne**
#### Summary

This pull request aims to centralize the Go version settings in the Mimir project. To achieve this we retrieve the Go version always from mimir-build-image.

#### Details:

Previously, the Go version setting was scattered across different parts of the Mimir project, making it difficult to manage and update consistently. By centralizing the Go version definition in the Makefile, we establish a single source of truth for the project's Go version.

This change simplifies the process of updating the Go version in the future, as it can now be modified in just one place. Additionally, any component or module that needs to reference the Go version can retrieve it from the centralized location defined in the Makefile.

The next step could be implementing a linter to enforce the rule of forbidding the use of the Go version outside dockerfile of mimir-build-image. This linter will scan the project files when PR and report any violations, ensuring that the Go version is only used in the designated place.

Please review this pull request and provide your feedback.

#### Checklist

- n/a Tests updated
- n/a Documentation added
- n/a `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by charleskorn — COMMENTED** [review]
Rather than moving the Go version into the Makefile, could we instead leave it as-is in the Dockerfile and run `go version` inside the build image when we want to retrieve the Go version?

This would mean we could continue to use tooling like Dependabot to automatically update the Go version. With the changes proposed here, we'd have to either manually update the Makefile every time we want to upgrade, or create our own tooling to update it automatically.

**Comment by ying-jeanne** [comment]
> 
Hi Charles, thank you for your input. I understand your suggestion. While in this PR I would actually centralizing the Alpine/Go version dependency to a single location, it would greatly simplify the update process. As the next step, I propose that if there is an update to the go/alpine version in the Makefile and the pull request is tagged as "release-needed," we can automate the release of the build image and Mimir. By automating this process, we can ensure a smoother and more efficient workflow.
If you have any other ideas or suggestions on how we can effectively centralize the version in a single location, I would greatly appreciate your input. Let's work together to find the best approach.

**Comment by andyasp** [comment]
I want to take a step back here. As I understand it the current situation is as follows:
1. Updating the Go version is a chore in this project.
2. We have a build image, which complicates it a bit.
3. We want to make it easier.
5. If it's easy enough maybe some of it can be automated.

Let's look at some of the recent pull requests where people actually had to confront this:

-  In https://github.com/grafana/mimir/pull/4904 it looks like it went pretty smoothly. What catches my eye is the `dev.dockerfile` files are using `golang:1.20.4` and not `golang:1.20.4-bullseye`. The build image tag also has random branch names in it, rather than a descriptive tag like what go version it is or a semver.
- Next there was https://github.com/grafana/mimir/pull/5184 which added a Make command to print the build image tag, so it can be used to replace the tag being specified in a workflow file. Seems fine, less room for error and less to review.
- The next Go update, https://github.com/grafana/mimir/pull/5185, still had to update a workflow file with the build image tag in `.github/workflows/helm-ci.yml` (should that be fixed?). Review caught some modifications that were needed.

Now we are here, where the tag is printed from the Makefile to access the build image to print (by running in the build image) the go version to define it in the workfile. In my opinion this is starting to stray from simplicity.

I have a couple of suggestions:
- Can there be a make command that does what the [docs](https://github.com/grafana/mimir/blob/main/docs/internal/contributing/how-to-upgrade-golang-version.md) say to do? I assume the reason this is being avoided is out of fear that the scripts will break, but a sed command on Dockerfiles/dev.dockerfile/workflow files for `golang:1.20.4-bullseye` and `mimir/build-image:go-1.20.4-blah` doesn't seem very fragile to me. Using a build argument for the base image came up as an issue in https://github.com/grafana/mimir/pull/5254.
- If I'm alone in the above sentiment and we want to continue in this direction, can the build image tag include the go version being used (along with something else, to allow patching) and then derive the go version from that tag? We're in control of it.

**Comment by charleskorn** [comment]
What if we had a workflow like this?

* Dependabot is configured to update the base image for the build image, but not the other Dockerfiles (otherwise it would create separate PRs for each Dockerfile)

* On PRs that update the build image, we have a GitHub Actions job that automatically triggers and runs the script that @andyasp suggested above. 

  * This job updates all other Dockerfiles and the GitHub Actions workflow files to keep them in sync and pushes the change to the PR branch. 
  * We could even have this run for all PRs so we can update the base image manually if we want to.

* On PRs that update the build image, we have a second GitHub Actions job that automatically builds and pushes an updated build image to Docker Hub. 

  * We probably want some kind of maintainer approval so that random people can't raise PRs that publish images to the `grafana` Docker Hub org.
  * If we don't have a deterministic way to determine the expected build image tag based on the state of the repo (ie. we need to hardcode it in the Makefile), then once the build image is pushed to Docker Hub, this job could also add a commit to the PR to update the build image tag to match.

(side note: this is way easier to do with Renovate instead of Dependabot, we can configure it to update all Docker images and the workflow files in one go rather than needing the first GitHub Actions job - I suspect using Renovate is not a possibility, but it's cases like these where Dependabot becomes a bit of a pain)

**Comment by pstibrany** [comment]
> What if we had a workflow like this?
> 
> * Dependabot is configured to update the base image for the build image, but not the other Dockerfiles (otherwise it would create separate PRs for each Dockerfile)

We would like to use Dependabot to update all Dockerfiles, but modify Dockerfiles so that only build-image depends on Go version. Other Dockerfiles that require Go version today will instead use build image. That's what this PR does.

> * On PRs that update the build image, we have a GitHub Actions job that automatically triggers and runs the script that @andyasp suggested above.

Yes, the plan is to automate building of the build-image via GitHub action, when its Dockerfile is updated.

>   * This job updates all other Dockerfiles and the GitHub Actions workflow files to keep them in sync and pushes the change to the PR branch.

With this PR we avoid the need for updating other Dockerfiles by using build image version from Makefile. In the next step we would like to autocompute build-image tag, and then the need for updating Makefile will go away too.

> * We could even have this run for all PRs so we can update the base image manually if we want to.
> * On PRs that update the build image, we have a second GitHub Actions job that automatically builds and pushes an updated build image to Docker Hub.

👍 Yes, see above.

>   * We probably want some kind of maintainer approval so that random people can't raise PRs that publish images to the `grafana` Docker Hub org.

That's a good point.

>   * If we don't have a deterministic way to determine the expected build image tag based on the state of the repo

We plan to introduce that.

**Comment by ying-jeanne** [comment]
I will copy discussion with Marco and Peter yesterday here:  Based on our discussions, we have agreed upon the following approach:

- Dependabot for Alpine updates: We will utilize Dependabot to handle the automatic updates of the Alpine base image. This ensures that we stay up to date with the latest Alpine version, benefiting from its security patches and improvements.

- Single point of truth for Go version: The single point of truth for the Go version will be maintained in mimir-build-image. This means that mimir-build-image will serve as the reference for the Go version used in our project.

- No single point of truth for Alpine: Unlike the Go version, we won't enforce a single point of truth for the Alpine base image.

- Image rebuild triggered when mimir-build-image Dockerfile changed: The process of rebuilding images will be triggered when dockerfile changed and PR approved. Upon triggering the rebuild, the associated Dockerfile will be tagged with a unique hash. and append a commit with new build-image into PR.

This is the plan that Peter mentioned above

**Comment by pstibrany** [comment]
> In my opinion this is starting to stray from simplicity.

Do you think it's going too far? This PR will make build-image the single source of truth for which Go version we use. I'm not sure if it's the best place, but at least it's a single place.

> I have a couple of suggestions:
> 
> * Can there be a make command that does what the [docs](https://github.com/grafana/mimir/blob/main/docs/internal/contributing/how-to-upgrade-golang-version.md) say to do? I assume the reason this is being avoided is out of fear that the scripts will break, but a sed command on Dockerfiles/dev.dockerfile/workflow files for `golang:1.20.4-bullseye` and `mimir-build-image:go-1.20.4-blah` doesn't seem very fragile to me.

If we can not have the extra script, isn't that better?

> Using a build argument for the base image came up as an issue in [Chore: Remove hardcoded Alpine Version in Dockerfiles #5254](https://github.com/grafana/mimir/pull/5254).

What problem do you see? I understand from your comment that you prefer templating, but that's just another way of achieving the same. If we can use docker build argument, why not use it?

> * If I'm alone in the above sentiment and we want to continue in this direction, can the build image tag include the go version being used (along with something else, to allow patching) and then derive the go version from that tag? We're in control of it.

We plan to make build-image tag to be computed as a hash of Dockerfile. I'm fine if we include golang version into it too, but would not try to parse the Go version from the tag. Instead, we could put it into image label and get it by running `docker inspect -f '{{ index .Config.Labels "label_name" }}' <build_image>`.

**Comment by pstibrany** [comment]
> We plan to make build-image tag to be computed as a hash of Dockerfile.

Idea here is to remove the need of updating Makefile with used tag name.

**Comment by andyasp** [comment]
> What if we had a workflow like this?

This overall flow makes sense to me, if the build-image changed then it looks like it's time to push a new image. I very much agree that we need a human in the loop before any image push happens. 

> Do you think it's going too far? This PR will make build-image the single source of truth for which Go version we use. I'm not sure if it's the best place, but at least it's a single place.

It feels like one step too far. If others disagree I'm fine with that, just expressing my view. For instance, I'm fine with `make print-build-image`, it's an `echo`, all well and good. Building off of that to create another layer of indirection makes me worry a bit where this is all going.

I'll give a small silly example to try to illustrate where my concern is coming from. There's a YAML parsing gotcha where an unquoted `1.20` will resolve to `"1.2"` while specifying a go version in a Github Action (here's a random [post](https://xeiaso.net/blog/go-1.20-yaml) for reference).

Now here is the diff in `github/workflows/test-build-deploy.yml`:
```diff
-          go-version: 1.20.5
+          go-version: ${{ needs.goversion.outputs.version }}
```

I guess we were lucky we specified a patch version before. I see where the idea of having a single source is coming from because `github/workflows/compare-helm-with-jsonnet.yml` does quote the version. But what's the situation now?
Indirection is obscuring auditability here and it's already a place hardly anyone looks at frequently.

If you trust and understand this end-to-end then it might feel solved, if you don't then it's a barrier to seeing what is actually happening.

> If we can not have the extra script, isn't that better?

It depends on the cost of not having the script.
 
> What problem do you see? I understand from your comment that you prefer templating, but that's just another way of achieving the same. If we can use docker build argument, why not use it?

The difference is the loss of visibility of the end result. Seeing rendered changes is valuable for catching mistakes.

> We plan to make build-image tag to be computed as a hash of Dockerfile.

There was a pretty lengthy discussion about something similar in an internal PR [here](https://github.com/grafana/backend-enterprise/pull/4657#discussion_r1202751473) you might be interested in. I'm starting to feel like a Luddite!

File: development/mimir-microservices-mode/dev.dockerfile
Comment (charleskorn) [inline development/mimir-microservices-mode/dev.dockerfile:2]: Do we need the default value here? A comment might be clearer.
```suggestion
ARG BUILD_IMAGE # Use ./compose-up.sh to build this image.
```

File: development/mimir-microservices-mode/dev.dockerfile
Comment (charleskorn) [inline development/mimir-microservices-mode/dev.dockerfile:3]: The braces aren't necessary.

```suggestion
FROM $BUILD_IMAGE
```

File: development/mimir-read-write-mode/dev.dockerfile
Comment (charleskorn) [inline development/mimir-read-write-mode/dev.dockerfile:1]: Same comments apply here as for `development/mimir-microservices-mode/dev.dockerfile`

**Review by charleskorn — APPROVED** [review]
Thanks for the extra context @pstibrany and @ying-jeanne. I'm OK with trying this approach, we can always change it later 🙂

File: .github/workflows/compare-helm-with-jsonnet.yml
Comment (ying-jeanne) [inline .github/workflows/compare-helm-with-jsonnet.yml:33]: here we can't really run make print-go-version, knowing in github workflow, spin up a docker image inside is not really supported, we need to run mimir-build-image to get the go version. and it is what is doing here. but there is no make target in mimir-build-image itself.

File: .github/workflows/test-build-deploy.yml
Comment (ying-jeanne) [reply .github/workflows/test-build-deploy.yml:43]: here we can't really run make print-go-version, knowing in github workflow, spin up a docker image inside is not really supported, we need to run container: image - mimir-build-image to get the go version. and it is what is doing here. but there is no make target in mimir-build-image itself.

**Review by pstibrany — APPROVED** [review]

File: Makefile
Comment (ying-jeanne) [reply Makefile:326]: it is like the lint, when outside of container it would just use local linter. we can try if you really think it is blocking.

**Review by pstibrany — APPROVED** [review]
Thanks, let's merge and learn how it works in practice.

File: Makefile
Comment (ying-jeanne) [reply Makefile:326]: this is working as `make lint` now, the default of make print-go-version is using mimir-build-image and when BUILD_IN_CONTAINER=false, it use local go version. Agreed we can revisit it later if it causes problem 👍 .

**Pull request merged**
